packages/runtime/src/pixelbuffer.ts

```diff
@@ -0,0 +1,269 @@
+     * @param pixeloffset
+     * @param r
+     * @param g
+     * @param b
+     */
+    setRgbAt(pixeloffset: number, r: number, g: number, b: number) {
+        const i = (this.start + (pixeloffset | 0)) * 3
+        this.buffer[i] = r
+        this.buffer[i + 1] = g
+        this.buffer[i + 2] = b
+    }
+
+    /**
+     * Reads the pixel color at the given offsret
+     * @param pixeloffset pixel offset. if negative starts from the end
+     * @returns
+     */
+    at(pixeloffset: number): number {
+        pixeloffset = pixeloffset | 0
+        while (pixeloffset < 0) pixeloffset += this.length
+        const i = this.start + pixeloffset
+        if (i < this.start || i >= this.start + this.length) return undefined
+        const bi = i * 3
+        const r = this.buffer[bi]
+        const g = this.buffer[bi + 1]
+        const b = this.buffer[bi + 2]
+        return rgb(r, g, b)
+    }
+
+    /**
+     * Apply a conversion function to all pixels
+     * @param converter
+     */
+    apply(converter: (c: number, index: number) => number) {
+        const n = this.length
+        const buf = this.buffer
+        for (let i = 0; i < n; ++i) {
+            const bi = (this.start + i) * 3
+
+            const r = buf[bi]
+            const g = buf[bi + 1]
+            const b = buf[bi + 2]
+            const c = rgb(r, g, b)
+
+            const cr = converter(c, i)
+
+            buf[bi] = (cr >> 16) & 0xff
+            buf[bi + 1] = (cr >> 8) & 0xff
+            buf[bi + 2] = cr & 0xff
+        }
+    }
+
+    /**
+     * Allocates a clone of the buffer view
+     * @returns
+     */
+    allocClone() {
+        const res = new PixelBuffer(
+            this.buffer.slice(this.start * 3, (this.start + this.length) * 3),
+            0,
+            this.length
+        )
+        return res
+    }
+
+    /**
+     * Clears the buffer to #000000
+     */
+    clear() {
+        this.buffer.fillAt(this.start * 3, this.length * 3, 0)
+    }
+
+    /**
+     * Creates a range view over the color buffer
+     * @param start start index
+     * @param length length of the range
+     * @returns a view of the color buffer
+     */
+    view(start: number, length?: number): PixelBuffer {
+        const rangeStart = this.start + (start << 0)
+        const rangeLength =
+            length === undefined
+                ? this.length - start
+                : Math.min(length, this.length - start)
+        return new PixelBuffer(this.buffer, rangeStart, rangeLength)
+    }
+}
+
+/**
+ * Create a color buffer that allows you to manipulate a range of pixels
+ * @param numPixels number of pixels
+ */
+export function pixelBuffer(numPixels: number) {
+    numPixels = numPixels | 0
+    const buf = ds.Buffer.alloc(numPixels * 3)
+    return new PixelBuffer(buf, 0, numPixels)
+}
+
+/**
+ * Writes a gradient between the two colors.
+ * @param startColor
+ * @param endColor
+ */
+export function fillGradient(
```

Comment:
I think these fills and rainbows should be methods on the PixelBuffer, otherwise they are rather hard to discover

Reply:
I'm trying to avoid to turn pixelbuffer into a monster